File: examples/rust/stateful-wots-gate/StatefulWOTSGate.runar.rs
use runar::prelude::*;

// StatefulWOTSGate - stateful + post-quantum interaction fixture (GAP-407).
#[runar::contract]
struct StatefulWOTSGate {
    count: Int,
}

impl StatefulWOTSGate {
    pub fn advance(&mut self, msg: &ByteString, sig: &ByteString, wots_pub_key: &ByteString) {
        assert!(verify_wots(msg, sig, wots_pub_key));
        self.count = self.count + 1;
    }
}

File: examples/zig/stateful-wots-gate/StatefulWOTSGate.runar.zig
const runar = @import("runar");

// StatefulWOTSGate — stateful + post-quantum interaction fixture (GAP-407).
pub const StatefulWOTSGate = struct {
    pub const Contract = runar.StatefulSmartContract;
    count: i64 = 0,

    pub fn init(count: i64) StatefulWOTSGate {
        return .{ .count = count };
    }

    pub fn advance(self: *StatefulWOTSGate, msg: runar.ByteString, sig: runar.ByteString, wotsPubKey: runar.ByteString) void {
        runar.assert(runar.verifyWOTS(msg, sig, wotsPubKey));
        self.count = self.count + 1;
    }
};
